doc/man-pages/man8/fms.8
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FMS 8"
.TH FMS 8 "2010-02-11" "OpenAFS" "AFS Command Reference"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
fms \- Determine a tape's capacity and a tape device's filemark size
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBfms\fR \fB\-tape\fR\ <\fItape\ special\ file\fR> [\fB\-help\fR]
.PP
\&\fBfms\fR \fB\-t\fR\ <\fItape\ special\ file\fR> [\fB\-h\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBfms\fR command determines the capacity of the tape currently in the
tape device identified by the \fB\-tape\fR argument, along with the size of
the filemark for the device. The filemark is also referred to as the
device's end-of-file (\s-1EOF\s0) marker, and can differ for each combination of
tape and tape device.
.PP
As the Tape Coordinator writes a dump, it writes a filemark between the
data included from each volume and also tracks the amount of space left
before the end of the tape (\s-1EOT\s0). For some tape devices, the filemark is
large enough (multiple megabytes) that failure to consider it leads the
Tape Coordinator significantly to overestimate the available space.
.PP
The intended use of this command is to determine tape capacity and
filemark size values that can be specified in a tape device's entry in the
\&\fI/usr/afs/backup/tapeconfig\fR file. For certain types of tape drives, the
Tape Coordinator operates more efficiently when the \fItapeconfig\fR file
lists accurate values. For further discussion, see the \fI\s-1IBM\s0 \s-1AFS\s0
Administration Guide\fR chapter on configuring the Backup System.
.PP
Insert a tape in the drive before issuing this command.
.SH "CAUTIONS"
.IX Header "CAUTIONS"
Do not use this command on compressing tape devices in compression mode or
with tape devices that handle tapes of multigigabyte (or multiterabyte)
capacity. It does not produce accurate results in those cases.  For
alternate suggestions on the values to record in the \fBtapeconfig\fR file
for compressing drives, see the \fI\s-1IBM\s0 \s-1AFS\s0 Administration Guide\fR chapter on
configuring the Backup System.
.PP
Running the command completely overwrites the tape, so use a blank one or
one that can be recycled.
.PP
Because it writes filemarks to the complete length of the tape, the
command can take from several hours to more than a day to complete.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-tape\fR <\fItape special file\fR>" 4
.IX Item "-tape <tape special file>"
Specifies the \s-1UNIX\s0 device name of the tape device for which to determine
filemark size and the capacity of the tape it currently contains. The
format varies on different system types, but usually begins with \fI/dev\fR;
an example is \fI/dev/sd0a\fR.
.IP "\fB\-help\fR" 4
.IX Item "-help"
Prints the online help for this command. All other valid options are
ignored.
.SH "OUTPUT"
.IX Header "OUTPUT"
The command generates output both on the standard output stream and in the
\&\fIfms.log\fR file that it creates in the current working directory. The
output reports the capacity of the tape in the device and the device's
filemark size.
.PP
The first few lines of output include status information about the
execution of the command, including such information as the number of
blocks and the number of file marks written to the tape by the
command. The last two lines of both screen and file output provide the
following information:
.IP "\(bu" 4
\&\f(CW\*(C`Tape capacity is \f(CInumber\f(CW bytes\*(C'\fR: specifies the size, in bytes, of the
tape in the device.
.IP "\(bu" 4
\&\f(CW\*(C`File marks are \f(CInumber\f(CW bytes\*(C'\fR: specifies the device's filemark size in
bytes.
.PP
The following message indicates that the fms command interpreter cannot
access the tape device. The command halts.
.PP
.Vb 1
\&   Can\*(Aqt open tape drive I<device>
.Ve
.PP
The following message indicates that the command interpreter cannot create
the \fIfms.log\fR log file. Again, the command halts.
.PP
.Vb 1
\&   Can\*(Aqt open log file
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following command illustrates the output for the device called
\&\fI/dev/rmt1h\fR:
.PP
.Vb 7
\&   % fms /dev/rmt1h
\&   wrote block: 130408
\&   Finished data capacity test \- rewinding
\&   wrote 1109 blocks, 1109 file marks
\&   Finished file mark test
\&   Tape capacity is 2136604672 bytes
\&   File marks are 1910205 bytes
.Ve
.PP
The following appears in the \fIfms.log\fR file:
.PP
.Vb 5
\&   fms test started
\&   wrote 9230 blocks
\&   Finished file mark test
\&   Tape capacity is 151224320 bytes
\&   File marks are 2375680 bytes
.Ve
.SH "PRIVILEGE REQUIRED"
.IX Header "PRIVILEGE REQUIRED"
The issuer must be able to insert and write to files in the currently
working directory, if the \fIfms.log\fR file does not already exist. If it
already exists, the issuer need only be able to write to it.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIfms.log\fR\|(5),
\&\fItapeconfig\fR\|(5)
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
\&\s-1IBM\s0 Corporation 2000. <http://www.ibm.com/> All Rights Reserved.
.PP
This documentation is covered by the \s-1IBM\s0 Public License Version 1.0.  It was
converted from \s-1HTML\s0 to \s-1POD\s0 by software written by Chas Williams and Russ
Allbery, based on work by Alf Wachsmann and Elizabeth Cassell.


NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.

* Changes incorporated in OpenAFS 1.2.5

** A remote denial of service attack in the AIX and IRIX clients has
   been fixed.  Users of those platforms are strongly encouraged to
   upgrade.

** Fixed race conditions in fileserver that could result in crash.

* Changes incorporated in OpenAFS 1.2.4

** Server logfiles now more consistant about format in which hosts are
   referred to.

** vfsck on Solaris will now allow force runs (using -y flag) even if old 
   inodes exist.

* Changes incorporated in OpenAFS 1.2.3

** Cell aliases for dynroot can be specified in the CellAlias file in
   /usr/vice/etc or /usr/local/etc/openafs, in format "realname alias",
   one per line.  They can also be managed at runtime with "fs newalias"
   and "fs listaliases".

* Changes incorporated in OpenAFS 1.2.2

** Solaris 9 and Linux PA-RISC are now supported

** fileserver will not erroneously delay legitimate errors for 3 seconds
   after 10 errors are returned (e.g. stat() on a directory you can't read)

** Rx MTU calculation now works for Irix, Solaris and Linux

** If afsd is started with the -dynroot flag, /afs will be locally
   generated from the CellServDB.  AFSDB cells will be mounted
   automatically upon access.

** The namei fileserver allows vice "partitions" to be directories instead
   of partitions and will attach and display accordingly. Creating the file
   "AlwaysAttach" in the /vicepX directory is used as the trigger to attach it.

** TSM support for butc no longer requires editing a Makefile, simply
   specify the --enable-tivoli-tsm configure option.

** Linux builds no longer require source changes every time the kernel

src/bu_utils/fms.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

#include <afsconfig.h>
#include <afs/param.h>


#undef	IN
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>		/* for mtio.h */
#include <afs/cmd.h>
#include <afs/procmgmt.h>
#include <afs/usd.h>

/* structure for writing data to tape */
typedef struct tapeDataBuffer {
    struct tapeDataBuffer *tdb_next;
    char *tdb_buffer;
} tapeDataBufferT;
typedef tapeDataBufferT *tapeDataBufferP;

/* globals */
char *tapeDevice = 0;		/* device pathname */
afs_int32 eotEnabled = 1;

/* prototypes */
int fileMark(usd_handle_t hTape);
int fileMarkSize(char *tapeDevice);
static int tt_fileMarkSize(struct cmd_syndesc *as, void *arock);
afs_int32 rewindTape(usd_handle_t hTape);
int dataBlock(usd_handle_t, afs_int32);

#define ERROR(evalue)                                           \
        {                                                       \
            code = evalue;                                      \
            goto error_exit;                                    \
        }

#define MAXV	100

#ifndef AFS_NT40_ENV
#include "AFS_component_version_number.c"
#endif

void quitFms(int);

int
main(int argc, char **argv)
{
    struct sigaction intaction, oldaction;
    struct cmd_syndesc *cptr;

    memset(&intaction, 0, sizeof(intaction));
    intaction.sa_handler = quitFms;

    sigaction(SIGINT, &intaction, &oldaction);

    cptr =
	cmd_CreateSyntax(NULL, tt_fileMarkSize, NULL,
			 "write a tape full of file marks");
    cmd_AddParm(cptr, "-tape", CMD_SINGLE, CMD_REQUIRED, "tape special file");

    cmd_Dispatch(argc, argv);
    return 0;
}

static int
tt_fileMarkSize(struct cmd_syndesc *as, void *arock)
{
    char *tapeDevice;

    tapeDevice = as->parms[0].items->data;

    fileMarkSize(tapeDevice);

    return 0;
}


int
fileMarkSize(char *tapeDevice)
{
    afs_uint32 nFileMarks, nBlocks, nbfTape;
    double tpSize, fmSize;
    afs_uint32 bufferSize = 16384;
    usd_handle_t hTape;
    FILE *logFile;
    int count = 0;
    afs_uint32 countr;
    afs_int32 code = 0;

    code =
	usd_Open(tapeDevice, (USD_OPEN_RDWR | USD_OPEN_WLOCK), 0777, &hTape);
    if (code) {
	printf("Can't open tape device %s\n", tapeDevice);
	fflush(stdout);
	exit(1);
    }

    logFile = fopen("fms.log", "w+");
    if (logFile == NULL) {
	printf("Can't open log file\n");
	fflush(stdout);
	exit(1);
    }
    fprintf(logFile, "fms test started\n");
    fflush(logFile);

    code = rewindTape(hTape);
    if (code) {
	fprintf(logFile, "Can't rewind tape\n");
	fflush(logFile);
	ERROR(code);
    }

    /* measure capacity of tape */
    nbfTape = 0;
    countr = 0;
    while (1) {
	code = dataBlock(hTape, bufferSize);
	nbfTape++;
	count++;
	countr++;
	if (code)
	    break;

	if (count >= 5) {
	    count = 0;
	    printf("\rwrote block: %d", nbfTape);
	}

    }

    fprintf(logFile, "wrote %d blocks\n", nbfTape);
    fflush(logFile);
    printf("\rwrote %d blocks\n", nbfTape);
    printf("Finished data capacity test - rewinding\n");
    /* reset the tape device */
    code = USD_CLOSE(hTape);
    if (code) {
	fprintf(logFile, "Can't close tape device at end of pass 1\n");
	fflush(logFile);
	printf("Can't close tape device %s\n", tapeDevice);
	goto error_exit;
    }
    code =
	usd_Open(tapeDevice, (USD_OPEN_RDWR | USD_OPEN_WLOCK), 0777, &hTape);
    if (code) {
	fprintf(logFile, "Can't open tape device for pass 2\n");
	fflush(logFile);
	printf("Can't open tape device %s\n", tapeDevice);
	goto error_exit;
    }

    code = rewindTape(hTape);
    if (code) {
	fprintf(logFile, "Can't rewind tape\n");
	fflush(logFile);
	ERROR(code);
    }

    /* now measure file mark size */
    nFileMarks = 0;
    nBlocks = 0;
    count = 0;
    countr = 0;
    while (1) {
	code = dataBlock(hTape, bufferSize);
	nBlocks++;
	if (code)
	    break;
	code = fileMark(hTape);
	nFileMarks++;
	if (code)
	    break;
	count++;
	countr++;

	if (count >= 2) {
	    count = 0;
	    printf("\rwrote %d blocks, %d filemarks", nBlocks, nFileMarks);
	}

    }
    printf("\nFinished filemark test\n");
    tpSize = (double)nbfTape *(double)bufferSize;
    fmSize =
	(((double)nbfTape -
	  (double)nBlocks) * (double)bufferSize) / (double)nFileMarks;
    printf("Tape capacity is %.0f bytes\n", tpSize);
    printf("File marks are %.0f bytes\n", fmSize);
    fprintf(logFile, "Tape capacity is %.0f bytes\n", tpSize);
    fprintf(logFile, "File marks are %.0f bytes\n", fmSize);
    fflush(logFile);
    fclose(logFile);
  error_exit:
    USD_CLOSE(hTape);
    return (code);
}

void
quitFms(int sig)
{
    exit(0);
}


/* --------------------------
 * device handling routines
 * --------------------------
 */

/* rewindTape() - rewinds tape to beginning */
afs_int32
rewindTape(usd_handle_t hTape)
{
    usd_tapeop_t tapeop;
    int rcode;

    tapeop.tp_op = USDTAPE_REW;
    tapeop.tp_count = 1;
    rcode = USD_IOCTL(hTape, USD_IOCTL_TAPEOPERATION, (void *)&tapeop);
    return rcode;
}

/* write an EOF marker */
int
fileMark(usd_handle_t hTape)
{
    usd_tapeop_t tapeop;
    int rcode;

    tapeop.tp_op = USDTAPE_WEOF;
    tapeop.tp_count = 1;
    rcode = USD_IOCTL(hTape, USD_IOCTL_TAPEOPERATION, (void *)&tapeop);
    return rcode;
}

/* dataBlock
 *	write a block of data on tape
 * entry:
 * 	blocksize - size of block in bytes
 */
int
dataBlock(usd_handle_t hTape, afs_int32 reqSize)
{
    static char *dB_buffer = 0;
    static afs_int32 dB_buffersize = 0;
    static int dB_count = 0;
    int *ptr;
    afs_int32 code = 0;
    afs_uint32 xferd;

    /* dbBuffersize is only valid when dB_buffer is non-zero */

    if ((dB_buffer != 0)
	&& (dB_buffersize != reqSize)
	) {
	free(dB_buffer);
	dB_buffer = 0;
    }

    if (dB_buffer == 0) {
	dB_buffer = (char *)malloc(reqSize);
	if (dB_buffer == 0)
	    ERROR(-1);
	dB_buffersize = reqSize;
	memset(dB_buffer, 0, dB_buffersize);
    }

    ptr = (int *)dB_buffer;
    *ptr = dB_count++;

    code = USD_WRITE(hTape, dB_buffer, dB_buffersize, &xferd);
    if (code || xferd != dB_buffersize)
	ERROR(-1);

  error_exit:
    return (code);
}
